dashboard/app/models/user.rb

```diff
@@ -1563,4 +1577,67 @@ def associate_with_potential_pd_enrollments
       end
     end
   end
+
+  private
+
+  # This method will extract a list of hidden ids by section. The type of ids depends
+  # on the input. If hidden_stages is true, id is expected to be a script id and
+  # we look for stages that are hidden. If hidden_stages is false, id is expected
+  # to be a course_id, and we look for hidden scripts.
+  # @param {string} assign_id - Course/script id we're asking for hidden elements of
+  # @param {boolean} [hidden_stages] - Whether we're looking for hidden stages or scripts
+  # @return {Hash<string,number[]>|number[]}
+  #   For teachers, this will be a hash mapping from section id to a list of hidden
+  #   ids for that section.
+  #   For students this will just be a list of ids that are hidden for them.
+  def get_hidden_ids(assign_id, hidden_stages = true)
```

Comment:
I didn't notice until this was moved to private that this method isn't called directly from outside.

It looks like the entry points are always explicitly through either `get_hidden_stage_ids` or `get_hidden_script_ids` so does it make sense to combine into one method just to split again based on the `hidden_stages` parameter?

Might it make more sense to keep the stage vs. script logic separate and converge on filtering the sections by user type and constructing the final array vs. hash?

Reply:
The main source of complexity here is in figuring out what set of sections to look at for students when determining which stages/scripts are hidden. As such, that's the bit I'd like to make sure I'm not repeating/introducing subtle differences to.

It might make sense however to get rid of `get_hidden_ids` and to have each of the callers instead just own that check, i.e.

```ruby
def get_hidden_stage_ids(script_name)
    script = Script.get_from_cache(script_name)
    return [] if script.nil?

    teacher? ? get_teacher_hidden_ids(true) : get_student_hidden_ids(script.id, true)
  end

  def get_hidden_script_ids(course)
    return [] if course.nil?

    teacher? ? get_teacher_hidden_ids(true) : get_student_hidden_ids(course.id, true)
  end
```

Thoughts?

Reply:
I presume you meant to pass `false` in the 2nd case?

I like it. That seems cleaner. 
Since this will still return either a hash or array, I'd like to see that `@returns` comment on each of these (public) methods.

Is there a chance we'll want types other than stage or script in the future? Should this be an enum (e.g. hidden_id_type) instead of a bool?

Reply:
Yep, meant to have false in second case.

I don't think we'll be wanting other types in the future.